proposals/kroxylicious-operator-api-alternative.md

```diff
@@ -0,0 +1,650 @@
+# Kroxylicious Operator API - Alternative Proposal.
+
+Keith isn't sure that 'Gateway API flavoured' approach for modelling ingress is the right one for Kroxylicious Operator.
+He thinks that our ingress API should work with Gateway APIs, rather than be one itself.  So, this 
+has prompted him to think about an alternative API design.
+
+## CRDs
+
+* Proxy CR - an instance of the Kroxylicious
+* ProxyIngress CR - Defines a way to access a Proxy
+* VirtualCluster CR - a virtual cluster
+* Filter CR - a filter definition
+
+![image](https://github.com/user-attachments/assets/8767da6b-3fc4-4fbc-99bc-64e93c06bc70)
+
+https://excalidraw.com/#json=CVAZMl-MuN1QP2cIW3hcA,cEWi5VjUg1mK2N1lvU-_xQ
+
+
+## Personas / Responsibilies
+
+* Infrastructure admin - responsible for the proxy tier and how proxies are exposed to the network.
+* Developer - responsible for the virtualcluster and filter
+
+## API
+
+### Proxy
+
+The Proxy CR is the responsibilty of the the Infrastructure admin.
+
+Declares an instance of the proxy and defines the ingress mechanisms available to it.
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: Proxy
+metadata:
+  name: myproxy
+spec:
+
+   # Optional - controls the Proxy infrastructure.
+   infrastructure:
+      # Influences the annotations and labels on all resources created by the operator.
+      annotations:
+         x: y
+      labels:
+         x: y
+      proxy:
+        # Controls specific to operand deployment
+        replicas: 2
+        resources:
+          requests:
+            memory: "64Mi"
+            cpu: "250m"
+#  # Optional: default ordered list of filters to be used by a virtualcluster that doesn't apply it own.  (this doesn't fit well with the Infra role - drop it?).
+#  defaultFilterRefs:
+#  - group: filter.kroxylicious.io
+#    kind: Filter
+#    name: encryption```
+
+status:
+   # status of the proxy tbd
+   conditions:
+   - status: x
+     type: y
+     reason: r
+   # per listener status 
+   ingresses:
+   - name: myclusterip
+     condidtions:
+     - ...
+```
+
+### ProxyIngress
```

Comment:
I'm tempted to call this a ProxyListener.

Reply:
I certainly think Listener is better than ingress, though does `listener` clash with Kafka listeners (for our user base)? Maybe `proxyAccess`

Reply:
Naming is so hard!

> I certainly think Listener is better than ingress, though does listener clash with Kafka listeners 

Isn't it the proxy's analogue for kafka's listener?   So maybe the clash is exactly what we want.

> Maybe proxyAccess

I think I prefer either Ingress or Listener over Access.  Access makes with think of ACLs.